### ISSUE TITLE
With prepend the aliased method gets the wrong self target

### ISSUE THREAD

**Issue by orhantoy** [jruby/jruby#6776]
I was seeing something very odd when using `awesome_print` with Rails 5.0+ on JRuby: this [aliased method](https://github.com/awesome-print/awesome_print/blob/8a7ff0aabacbebf694c3e27242809219a96d5a3b/lib/awesome_print/core_ext/string.rb#L29) had the class as `self`, not the instance itself.

@kares helped debug this and created this reproduction script:

```ruby
module ActiveSupport
  module ToJsonWithActiveSupportEncoder
  end
end

String.prepend(ActiveSupport::ToJsonWithActiveSupportEncoder)

class String

  i = 42

  define_method :whiteish do |*html|
    fail "given #{self.inspect} (#{self.class})" unless self.is_a?(String)
    "#{self}" # passes (when next line is commented)
    "#{1 + i} - #{self}"
  end

  alias :pale  :whiteish

end

p '=>'.whiteish
p '=>'.pale
```

**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc): latest 9.2.19.0
* Operating system and platform (e.g. `uname -a`): `Darwin Kernel Version 20.6.0`

**Expected Behavior**

`self` in the aliased method should be the instance itself, not the class. See above reproduction script.

**Actual Behavior**

`self` in the aliased method is the class _when_ `prepend` is used.

**Comment by orhantoy** [comment]
Here's a simpler reproduction script:

```ruby
module SomeModule
end

String.prepend(SomeModule)

class String
  i = 42

  define_method :foo do
    [i, self, self.class]
  end

  alias :bar :foo
end

p 'foo'.foo
p 'bar'.bar
```

JRuby output (incorrect):
```
[42, "foo", String]
[42, String, Class]
```

MRI output (expected):
```
[42, "foo", String]
[42, "bar", String]
```

**Comment by headius** [comment]
This is probably all still part of the same issue that @enebo has been investigating in #6445 #6271 #6347 and others.

**Comment by enebo** [comment]
This is not actually the same issue but since I have fixed the other 3 I will see how hard this one is.

**Comment by enebo** [comment]
The branch I am working on is for 9.3 and not 9.2.  I believe a one-line change will fix this there too...so I will leave this open so I don't forget to fix and verify that before our next 9.2 release.
